### PR TITLE
Optimize kv_cache for gpt-neox model family

### PR DESCRIPTION
**Pull request by hkvision**

### CODE REVIEW

**Review by lalalapotter — APPROVED** [review]
LGTM

**Review by sgwhat — APPROVED** [review]

File: python/llm/src/bigdl/llm/transformers/models/gptneox.py [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+
+
+def apply_rotary_pos_emb(q, k, cos, sin, position_ids):
+    gather_indices = position_ids[:, None, :, None]  # [bs, 1, seq_len, 1]
+    gather_indices = gather_indices.repeat(1, cos.shape[1], 1, cos.shape[3])
+    cos = torch.gather(cos.repeat(gather_indices.shape[0], 1, 1, 1), 2, gather_indices)
+    sin = torch.gather(sin.repeat(gather_indices.shape[0], 1, 1, 1), 2, gather_indices)
+    q_embed = (q * cos) + (rotate_half(q) * sin)
+    k_embed = (k * cos) + (rotate_half(k) * sin)
+    return q_embed, k_embed
```
[the remark this answers is not in the document]
Comment (hkvision) [reply python/llm/src/bigdl/llm/transformers/models/gptneox.py:46]: ok, gptneox uses a different rotary_pos_emb, will add it into utils

File: python/llm/src/bigdl/llm/transformers/models/utils.py [hunk 2]
```diff
@@ -68,6 +68,14 @@ def apply_rotary_pos_emb(q, k, cos, sin, position_ids, model_family):
         q_embed = (q * cos) + (rotate_half(q) * sin)
         k_embed = (k * cos) + (rotate_half(k) * sin)
         return q_embed, k_embed
+    elif model_family == "gpt_neox":
+        gather_indices = position_ids[:, None, :, None]  # [bs, 1, seq_len, 1]
+        gather_indices = gather_indices.repeat(1, cos.shape[1], 1, cos.shape[3])
+        cos = torch.gather(cos.repeat(gather_indices.shape[0], 1, 1, 1), 2, gather_indices)
+        sin = torch.gather(sin.repeat(gather_indices.shape[0], 1, 1, 1), 2, gather_indices)
+        q_embed = (q * cos) + (rotate_half(q) * sin)
+        k_embed = (k * cos) + (rotate_half(k) * sin)
```
Comment (yangw1234) [inline python/llm/src/bigdl/llm/transformers/models/utils.py:77]: These two implementations (gpt_neox and llama/baichuan) seems to be equivalent. see https://github.com/huggingface/transformers/commit/626c1b8af19f3d4cb2f2d2b116fdce45bed10ebb

Comment (hkvision) [reply python/llm/src/bigdl/llm/transformers/models/utils.py:77]: Correct, combined in a new PR, please take a look.

**Pull request merged**